File: src/lib/api/members.ts
import { invokeCommand } from "../tauri";

export interface MemberResponse {
  id: string;
  member_number: string;
  full_name: string;
  father_name: string | null;
  phone: string | null;
  cnic: string | null;
  address: string | null;
  date_of_birth: string | null;
  gender: string | null;
  notes: string | null;
  is_archived: boolean;
  membership_plan_name: string | null;
  membership_start_date: string | null;
  membership_expiry_date: string | null;
  membership_status: string | null;
  created_at: string;
  updated_at: string;
}

export interface CreateMemberRequest {
  full_name: string;
  father_name: string | null;
  phone: string | null;
  cnic: string | null;
  address: string | null;
  date_of_birth: string | null;
  gender: string | null;
  notes: string | null;
}

export interface UpdateMemberRequest {
  full_name: string;
  father_name: string | null;
  phone: string | null;
  cnic: string | null;
  address: string | null;
  date_of_birth: string | null;
  gender: string | null;
  notes: string | null;
}

export async function createMember(
  request: CreateMemberRequest,
): Promise<MemberResponse> {
  return invokeCommand<MemberResponse>("create_member", { request });
}

export async function getMember(id: string): Promise<MemberResponse> {
  return invokeCommand<MemberResponse>("get_member", { id });
}

export async function listMembers(args: {
  search?: string;
  status?: string;
  include_archived?: boolean;
}): Promise<MemberResponse[]> {
  return invokeCommand<MemberResponse[]>("list_members", {
    search: args.search ?? null,
    status: args.status ?? null,
    include_archived: args.include_archived ?? false,
  });
}

export async function updateMember(
  id: string,
  request: UpdateMemberRequest,
): Promise<MemberResponse> {
  return invokeCommand<MemberResponse>("update_member", { id, request });
}

export async function archiveMember(id: string): Promise<MemberResponse> {
  return invokeCommand<MemberResponse>("archive_member", { id });
}

File: src-tauri/src/commands/members.rs
use tauri::State;

use crate::database::Database;
use crate::dto::member::{CreateMemberRequest, MemberResponse, UpdateMemberRequest};
use crate::errors::AppError;
use crate::services::member_service;

use super::db::run_db;

#[tauri::command]
pub async fn create_member(
    state: State<'_, Database>,
    request: CreateMemberRequest,
) -> Result<MemberResponse, AppError> {
    let conn = state.inner().clone_conn();
    run_db(conn, move |c| member_service::create_member(c, request)).await
}

#[tauri::command]
pub async fn get_member(
    state: State<'_, Database>,
    id: String,
) -> Result<MemberResponse, AppError> {
    let conn = state.inner().clone_conn();
    run_db(conn, move |c| member_service::get_member(c, &id)).await
}

#[tauri::command]
pub async fn list_members(
    state: State<'_, Database>,
    search: Option<String>,
    status: Option<String>,
    include_archived: Option<bool>,
) -> Result<Vec<MemberResponse>, AppError> {
    let conn = state.inner().clone_conn();
    run_db(conn, move |c| {
        member_service::list_members(
            c,
            search.as_deref().unwrap_or(""),
            status.as_deref(),
            include_archived.unwrap_or(false),
        )
    })
    .await
}

#[tauri::command]
pub async fn update_member(
    state: State<'_, Database>,
    id: String,
    request: UpdateMemberRequest,
) -> Result<MemberResponse, AppError> {
    let conn = state.inner().clone_conn();
    run_db(conn, move |c| {
        member_service::update_member(c, &id, request)
    })
    .await
}

#[tauri::command]
pub async fn archive_member(
    state: State<'_, Database>,
    id: String,
) -> Result<MemberResponse, AppError> {
    let conn = state.inner().clone_conn();
    run_db(conn, move |c| member_service::archive_member(c, &id)).await
}

#[tauri::command]
pub async fn unarchive_member(
    state: State<'_, Database>,
    id: String,
) -> Result<MemberResponse, AppError> {
    let conn = state.inner().clone_conn();
    run_db(conn, move |c| member_service::unarchive_member(c, &id)).await
}
